docs: Async data (#5196)
dfinity/motoko: doc/md/5-actors-async.md

Instead of making the caller wait, the message is sent and queued, and the future is immediately returned. The caller can continue doing other work and later use `await` to pause until the future is ready and get the result.

You can also attach extra information to an async call using a prefixed parenthetical in the form `(base with attr₁ = v₁; attr₂ = v₂; …)` where `base` is an optional record containing (e.g., default) attributes. Accepted attributes are currently `cycles : Nat`, specifying the amount of cycles to be sent along with the message, and `timeout : Nat32` to modify the deadline and restrict the time span while the receiver can reply.

The combination of `async`/`await` constructs simplifies asynchronous programming  by allowing `await`s to be embedded within ordinary sequential code, without requiring tricky management of asynchronous callbacks.

## Traps and commit points

A trap is a non-recoverable runtime failure in Motoko, caused by errors such as:

* Division by zero.
* Out-of-bounds array access.
* Numeric overflow.
* Exceeding cycle limits.
* Failing an assertion.

> When a shared function executes without encountering an `await` expression, it does not suspend and runs atomically, meaning its execution cannot be interleaved or interrupted. Such functions are considered syntactically atomic because their structure guarantees uninterrupted execution from start to finish.

Reviewer: ```suggestion
When a shared function executes without evaluating an `await` expression, it never suspend and thus runs atomically, meaning its execution cannot be interleaved with that of another message. Functions that don't contain any `await` expressions are syntactically atomic and guarantee interference-free execution from start to finish.
```

Calling `nonAtomic()` also fails with an error, but the state is partially updated. In this case, `s` ends up as 3, and `pinged` is true. This happens because the first `await` commits all prior changes, including the message send. The second `await` causes another commit and suspends execution, allowing other messages to be processed before the trap occurs.

## Async functions

Here is an example program that uses async functions:

``` motoko file=../examples/counter-actor.mo
```

The `Counter` actor declares one field and three public, shared functions:

-   The field `count` is mutable, initialized to zero and implicitly `private`.

-   Function `inc()` asynchronously increments the counter and returns a future of type `async ()` for synchronization.

-   Function `read()` asynchronously reads the counter value and returns a future of type `async Nat` containing its value.

-   Function `bump()` asynchronously increments and reads the counter.

The only way to read or modify the state (`count`) of the `Counter` actor is through its shared functions.

## Using `await` to consume async futures

> The caller of a shared function typically receives a future, a value of type `async T` for some T.

Reviewer: ```suggestion
The caller of a shared function typically receives a future, a value of type `async T` for some `T`.
```

To access the result of an `async` value, the receiver of the future uses an `await` expression.

For example, to use the result of `Counter.read()` above, we can first bind the future to an identifier `a`, and then `await a` to retrieve the underlying [`Nat`](../base/Nat.md), `n`:

``` motoko no-repl
let a : async Nat = Counter.read();
let n : Nat = await a;
```

The first line immediately receives a future of the counter value, but does not wait for it, and thus cannot use it as a natural number yet.

The second line `await`s this future and extracts the result, a natural number. This line may suspend execution until the future has been completed.

Typically, one rolls the two steps into one and just awaits an asynchronous call directly:

``` motoko no-repl
let n : Nat = await Counter.read();
```

Unlike a local function call, which waits for the result before continuing, a shared function call returns a future immediately without blocking. Later, calling `await` on that future pauses the current task until the future is finished. When the future completes, `await` either returns the result or throws an error if the future ended with one.

> If you `await` the same future again, it just returns the same result or error. Even if the future is already done, `await` still pauses briefly to make sure all previous state changes and messages are saved before continuing.

Reviewer: ```suggestion
If you `await` the same future again, it just returns the same result or error. Even if the future is already done, `await` will briefly suspend to commit all pending state changes and message sends. This means that you can rely on every `await` to commit state, whether its future is still in progress or already completed.
```

Even with just a single async call, Motoko’s abstractions help keep the code clear. By using `await`, the programmer tells the compiler where interleaving can happen, avoiding the need to restructure logic to fit the system’s message-passing loop.

In other languages without these features, developers often need to use advanced patterns like callbacks and event handlers. This low-level, systems-style programming can be powerful but is error-prone, as it breaks high-level logic into scattered events and shared state.

### Example

To demonstrate how asynchronous actors work, consider the following example.

Customers place orders at a pizza restaurant, but the chef can only make one pizza at a time. Orders are taken **[asynchronously](https://internetcomputer.org/docs/motoko/fundamentals/actors-async#async--await)**, meaning customers do not have to wait for previous orders to be completed before placing their own. However, each pizza is prepared sequentially. This is representative of an asynchronous actor.

> 

Reviewer: ```suggestion
<!-- TODO: It would be cleaner to use a Deque or Queue pushing new order to the end and popping the next order to make from the front. -->
```

```motoko no-repl

> import Array "mo:base/Array";

Reviewer: It would be better use a Deque (Queue) pushing new order to the ends and popping the next order to make from the front.

When you use `await` on a future, it suspends the caller until the task finishes. But `await*` on a computation doesn’t suspend the caller; it immediately runs the computation like a normal function call.

This means that `await*` only causes suspension if the computation’s body itself uses a regular `await`. The `*` indicates that the computation might include zero or more `await` calls and so may be interleaved with other message executions.

You create an `async*` value by using an `async*` expression, but usually, it’s done by defining a local function that returns an `async*` type.

To get the result of an `async*` computation, you use `await*`.

:::danger

Use `async*` and `await*` carefully. In Motoko, a regular `await` is a commit point. State changes are saved before the function pauses.

`await*` is not a commit point because the computation it runs may not pause or commit at a predictable time. This means if a trap happens inside an `await*` computation, the actor’s state will roll back to the last commit point before the `await*`, not to the point of the `await*` itself.

:::

> ### Example
> 
> Customers can place multiple orders at the same time. However, when they ask for an update on their delivery status, responses must be generated sequentially based on the preparation time. Instead of constantly tracking delivery updates, `async*` can be used to calculate the estimated delivery time only when requested.
> 
> ```motoko no-repl
> // async* function to calculate delivery time on demand
> func checkDelivery(order : Text) : async* Text {
>         let estimatedTime = 15 + (orders.size() * 5); // Base time + 5 mins per pending order
>         return "Your " # order # " will arrive in " # Nat.toText(estimatedTime) # " minutes.";
>     };
> 
>     // Multiple users can check their delivery status asynchronously
>     public shared func getDeliveryStatus(order : Text) : async Text {
>         let status = await* checkDelivery(order);
>         return status;
>     };
> ```

Reviewer: ````suggestion
<!--
### Example

Customers can place multiple orders at the same time. However, when they ask for an update on their delivery status, responses must be generated sequentially based on the preparation time. Instead of constantly tracking delivery updates, `async*` can be used to calculate the estimated delivery time only when requested.

```motoko no-repl
// async* function to calculate delivery time on demand
func checkDelivery(order : Text) : async* Text {
        let estimatedTime = 15 + (orders.size() * 5); // Base time + 5 mins per pending order
        return "Your " # order # " will arrive in " # Nat.toText(estimatedTime) # " minutes.";
    };

    // Multiple users can check their delivery status asynchronously
    public shared func getDeliveryStatus(order : Text) : async Text {
        let status = await* checkDelivery(order);
        return status;
    };
```
-->
````

This is *not* a good example of using async*/await* since the checkDelivery function could just as well be written as a synchronous, local function - it doesn't do anything async in its body.
I'm pretty sure the original doc had a better example of this.

Author: Yes the async* info is in the section above (copied from that page and revised editorially, will use the old example here instead).

> ## `try/finally`

Reviewer: Are try-catch and try-catch-finally explained anywhere? There's used to be section on this in error handling but it seems to have disappeared.
